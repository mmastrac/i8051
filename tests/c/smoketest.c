/* test8051.c â€” SDCC 8051 test program */

#include <8051.h>  /* gives you SFR definitions like P1 */

void main(void)
{
    unsigned char i, sum = 0;
    __xdata unsigned char *scratch = (__xdata unsigned char *)0x8000;

    /* Clear scratchpad */
    for (i = 0; i < 4; i++) {
        scratch[i] = 0;
    }

    /* Simple loop: add numbers 1..10 into sum */
    for (i = 1; i <= 10; i++) {
        sum += i;
    }

    /* Store result in external memory for visibility */
    scratch[0] = sum;

    /* Expected sum of 1..10 = 55 decimal */
    if (sum == 55) {
        P1 = 0x55;   /* indicate success on Port 1 */
    } else {
        P1 = 0xAA;   /* indicate failure */
    }

    /* Stop here (endless loop) */
    while (1) {
        /* idle */
    }
}