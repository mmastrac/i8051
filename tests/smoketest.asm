; 8051 test case - output byte 1 to memory 0x8000

.module smoketest
.area CSEG (CODE)

_start:
    ; Set up DPTR to point to 0x8000
    mov DPTR, #0x8000

    ; Load accumulator with value 1
    mov A, #0x01

    ; Store accumulator to external memory at DPTR
    movx @DPTR, A

    ; Halt execution
halt:
    sjmp halt

.area DSEG (DATA)
.area XSEG (XDATA)
.area HOME (CODE)
.area PSEG (CODE)

