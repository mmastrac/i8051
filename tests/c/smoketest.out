P1 = 55
MEM: 37 00 00 00
